h2/src/main/org/h2/mvstore/db/MVTempResult.java

```diff
@@ -0,0 +1,170 @@
+import org.h2.util.TempFileDeleter;
+import org.h2.value.Value;
+
+/**
+ * Temporary result.
+ *
+ * <p>
+ * A separate MVStore in a temporary file is used for each result. The file is
+ * removed when this result and all its copies are closed.
+ * {@link TempFileDeleter} is also used to delete this file if results are not
+ * closed properly.
+ * </p>
+ */
+public abstract class MVTempResult implements ResultExternal {
+
+    /**
+     * Creates MVStore-based temporary result.
+     *
+     * @param session
+     *                        session
+     * @param expressions
+     *                        expressions
+     * @param distinct
+     *                        is output distinct
+     * @param sort
+     *                        sort order, or {@code null}
+     * @return temporary result
+     */
+    public static ResultExternal of(Session session, Expression[] expressions, boolean distinct, SortOrder sort) {
+        return distinct || sort != null ? new MVSortedTempResult(session, expressions, distinct, sort)
+                : new MVPlainTempResult(session, expressions);
+    }
+
+    /**
+     * MVStore.
+     */
+    final MVStore store;
+
+    /**
+     * Count of rows. Used only in a root results, copies always have 0 value.
+     */
+    int rowCount;
+
+    /**
+     * Parent store for copies. If {@code null} this result is a root result.
+     */
+    final MVTempResult parent;
+
+    /**
+     * Count of child results.
+     */
+    int childCount;
+
+    /**
+     * Whether this result is closed.
+     */
+    boolean closed;
+
+    /**
+     * Temporary file deleter.
+     */
+    private final TempFileDeleter tempFileDeleter;
+
+    /**
+     * Name of the temporary file.
+     */
+    private final String fileName;
+
+    /**
+     * Reference to the record in the temporary file deleter.
+     */
+    private final Reference<?> fileRef;
+
+    /**
+     * Creates a shallow copy of the result.
+     *
+     * @param parent
+     *                   parent result
+     */
+    MVTempResult(MVTempResult parent) {
+        this.parent = parent;
+        this.store = parent.store;
+        this.fileName = null;
+        this.tempFileDeleter = null;
+        this.fileRef = null;
+    }
+
+    /**
+     * Creates a new temporary result.
+     *
+     * @param session
+     *                    database session
+     */
+    MVTempResult(Session session) {
+        try {
+            fileName = FileUtils.createTempFile("h2tmp", Constants.SUFFIX_TEMP_FILE, false, true);
+            Builder builder = new MVStore.Builder().fileName(fileName);
+            byte[] key = session.getDatabase().getFileEncryptionKey();
+            if (key != null) {
+                builder.encryptionKey(MVTableEngine.decodePassword(key));
+            }
+            store = builder.open();
+            tempFileDeleter = session.getDatabase().getTempFileDeleter();
+            fileRef = tempFileDeleter.addFile(fileName, this);
```

Comment:
I'm not sure what to pass to `TempFileDeleter`. Probably its functionality is not very useful here. On Linux / UNIX, for example, we can usually remove the file even if it is in use, so on such systems this in not significant.

But Windows is different. We need to close storage somehow before deletion.

This problem is not very big, because this is a fallback code. Temporary results should be closed and closing sequence correctly closes the store, removes the file, and removes it from `TempFileDeleter` queue.

Reply:
I was wrong, this problem is significant. Results are not always closed in tests, file is deleted, but background writer thread is still alive and eat some memory.

Reply:
yeah, I suspect this is mostly to catch sessions that throw an exception somewhere deep in our code e.g. if the user has a dodgy custom trigger code

Reply:
I have an idea how to fix it. I'm trying to implement it now.